torch/numpy/__init__.py

```diff
@@ -0,0 +1,1604 @@
+
+import torch
+import torch.numpy.random
+
+
+def wrap_value(v):
+    if isinstance(v, torch.Tensor):
+        return TensorWrapper(v)
+    if type(v)==type and isinstance(object, v):
+        return None
+    else:
+         return v
+def unwrap_value(v):
+    if isinstance(v, TensorWrapper):
+        return v.tensor
+    if type(v)==type and isinstance(object, v):
+        return None
+    else:
+        return v
+
+
+def wrap_args(*args):
+    return [wrap_value(v) for v in args]
+def wrap_kwargs(**kwargs):
+    return dict([(k, wrap_value(kwargs[k])) for k in kwargs])
+
+def unwrap_args(*args):
+    return [unwrap_value(v) for v in args]
+def unwrap_kwargs(**kwargs):
+    result = dict([(k, unwrap_value(kwargs[k])) for k in kwargs])
+    return result
+
+#import numpy as orig_np
+class TensorWrapper:
+    tensor=0
+
+    def __init__(self, t):
+        self.tensor=t
+
+    @property
+    def dtype(self):
+        return self.tensor.dtype
+
+    @property
+    def ndim(self):
+        return self.tensor.ndim
+
+    def __abs__(*args, **kwargs):
+        return wrap_value(args[0].tensor.__abs__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    def __add__(*args, **kwargs):
+        return wrap_value(args[0].tensor.__add__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    def __and__(*args, **kwargs):
+        return wrap_value(args[0].tensor.__and__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    def __array__(*args, **kwargs):
+        return wrap_value(args[0].tensor.__array__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    def __array_finalize__(*args, **kwargs):
+        return wrap_value(args[0].tensor.__array_finalize__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    #def __array_interface__(*args, **kwargs):
+    #    return wrap_value(args[0].tensor.__array_interface__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    def __array_prepare__(*args, **kwargs):
+        return wrap_value(args[0].tensor.__array_prepare__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    def __array_priority__(*args, **kwargs):
+        return wrap_value(args[0].tensor.__array_priority__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    #def __array_struct__(*args, **kwargs):
+    #    return wrap_value(args[0].tensor.__array_struct__(*unwrap_args(*args[1:]), **unwrap_kwargs(**kwargs)))
+    def __array_ufunc__(*args, **kwargs):
```

Comment:
I'd expect you to need both `__array_function__` and `__array_ufunc__`, or neither. They do basically the same job, one for ufuncs and one for all other functions that take an `array_like` as input.